File: ruby/view/singlelineformatter.rb
# Created by Satoshi Nakagawa.
# You can redistribute it and/or modify it under the Ruby's license or the GPL2.

require 'utility'

class SingleLineFormatter < NSFormatter
  
  def stringForObjectValue(str)
    str.to_s.gsub(/\r\n|\r|\n/, ' ')
  end
  
  def getObjectValue_forString_errorDescription(objp, str, err)
    s = str.to_s.gsub(/\r\n|\r|\n/, ' ')
    objp.assign(s.to_ns)
    true
  end
  
  def isPartialStringValid_newEditingString_errorDescription(str, strp, err)
    s = str.to_s
    return true unless s =~ /\r\n|\r|\n/
    s = s.gsub(/\r\n|\r|\n/, ' ')
    strp.assign(s.to_ns)
    false
  end
end

File: objc/Views/SingleLineFormatter.m
// Created by Satoshi Nakagawa.
// You can redistribute it and/or modify it under the Ruby's license or the GPL2.

#import "SingleLineFormatter.h"


@implementation SingleLineFormatter

- (NSString*)stringForObjectValue:(id)str
{
	str = [str stringByReplacingOccurrencesOfString:@"\r" withString:@" "];
	return [str stringByReplacingOccurrencesOfString:@"\n" withString:@" "];
}

- (BOOL)getObjectValue:(id*)obj forString:(NSString*)str errorDescription:(NSString**)err
{
	str = [str stringByReplacingOccurrencesOfString:@"\r" withString:@" "];
	str = [str stringByReplacingOccurrencesOfString:@"\n" withString:@" "];
	*obj = str;
	return YES;
}

- (BOOL)isPartialStringValid:(NSString *)str newEditingString:(NSString **)newString errorDescription:(NSString **)err
{
	NSRange r = [str rangeOfCharacterFromSet:[NSCharacterSet newlineCharacterSet]];
	if (r.location == NSNotFound) return YES;
	
	str = [str stringByReplacingOccurrencesOfString:@"\r" withString:@" "];
	str = [str stringByReplacingOccurrencesOfString:@"\n" withString:@" "];
	*newString = str;
	return NO;
}

@end
